R/ppc-loo.R

```diff
@@ -456,3 +519,81 @@ ppc_loo_ribbon <-
   return(psis_object)
 }
 
+## Boundary correction based on code by Yang Hu and Carl Scarrott in evmix package
+
+# Boundary correction KDE helper function
+.bc_dunif <- function(xs, pvals, b, xmax = 1){
+  # Function based on biased-corrected (modified) beta kernel 
+  # Chen, Song Xi. "Beta kernel estimators for density functions." 
+  # Computational Statistics & Data Analysis 31.2 (1999): 131-145.
+  
+  # Re-scaling inputs based on upper_bound given beta kernel is defined [0,1]
+  xs <- xs/xmax
+  pvals <- pvals/xmax
+  b <-  b/xmax # smoothing parameter (i.e. bw)
+  
+  d <- vapply(X = xs, 
+              FUN = .bc_kde_calc, 
+              b = b,
+              pvals = pvals, 
+              FUN.VALUE = 0)
```

Comment:
Hello! I don't mean to chime in out of nowhere with a potentially useless suggestion, but one can usually get a reasonable performance bump by writing an `Rcpp` function that takes a `NumericVector` or `NumericMatrix` and does the looping / `vapply`ing inside `C++` instead of in `R` (at least in my own experience translating KDEs to `Rcpp`). 

It may be worth looking at something like [`sm::density`](https://github.com/cran/sm/blob/master/R/density.r), not for code (it's all in R) but for the methodology (make everything a vector/matrix and rely on the linear algebra operations as much as possible), as it is surprisingly speedy (sometimes) for a pure R implementation as a result.